# Week 3 — Data Preprocessing
**Internship:** IDX Exchange Data Science Program  
**Name:** Monika  
**Week:** 3  

**Goal:** Clean the data, select features, and create a train/test split ready for modeling. All decisions are documented with reasoning.

## 1. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

print('Libraries loaded!')

Libraries loaded!


## 2. Reload & Filter Data
Reloading all CSV files and filtering to Single Family Residences, same as Week 2.

In [4]:
data_folder = r'C:\Users\monik\OneDrive - University of Illinois - Urbana\Desktop\IDX Exchange_DS\data\california'

df1 = pd.read_csv(data_folder + r'\CRMLSSold20220101_20231231_filled.csv', low_memory=False)
df2 = pd.read_csv(data_folder + r'\CRMLSSold202401_filled.csv', low_memory=False)
df3 = pd.read_csv(data_folder + r'\CRMLSSold202402_filled.csv', low_memory=False)
df4 = pd.read_csv(data_folder + r'\CRMLSSold202403_filled.csv', low_memory=False)
df5 = pd.read_csv(data_folder + r'\CRMLSSold202404_filled.csv', low_memory=False)
df6 = pd.read_csv(data_folder + r'\CRMLSSold202405_filled.csv', low_memory=False)
df7 = pd.read_csv(data_folder + r'\CRMLSSold202406_filled.csv', low_memory=False)
df8 = pd.read_csv(data_folder + r'\CRMLSSold202407_filled.csv', low_memory=False)
df9 = pd.read_csv(data_folder + r'\CRMLSSold202408.csv', low_memory=False)
df10 = pd.read_csv(data_folder + r'\CRMLSSold202409.csv', low_memory=False)
df11 = pd.read_csv(data_folder + r'\CRMLSSold202410.csv', low_memory=False)
df12 = pd.read_csv(data_folder + r'\CRMLSSold202411.csv', low_memory=False)
df13 = pd.read_csv(data_folder + r'\CRMLSSold202412.csv', low_memory=False)
df14 = pd.read_csv(data_folder + r'\CRMLSSold202501_filled.csv', low_memory=False)
df15 = pd.read_csv(data_folder + r'\CRMLSSold202502.csv', low_memory=False)
df16 = pd.read_csv(data_folder + r'\CRMLSSold202503.csv', low_memory=False)
df17 = pd.read_csv(data_folder + r'\CRMLSSold202504.csv', low_memory=False)
df18 = pd.read_csv(data_folder + r'\CRMLSSold202505.csv', low_memory=False)
df19 = pd.read_csv(data_folder + r'\CRMLSSold202506.csv', low_memory=False)
df20 = pd.read_csv(data_folder + r'\CRMLSSold202507.csv', low_memory=False)
df21 = pd.read_csv(data_folder + r'\CRMLSSold202508.csv', low_memory=False)
df22 = pd.read_csv(data_folder + r'\CRMLSSold202509.csv', low_memory=False)
df23 = pd.read_csv(data_folder + r'\CRMLSSold202510.csv', low_memory=False)
df24 = pd.read_csv(data_folder + r'\CRMLSSold202511.csv', low_memory=False)
df25 = pd.read_csv(data_folder + r'\CRMLSSold202512.csv', low_memory=False)
df26 = pd.read_csv(data_folder + r'\CRMLSSold202601.csv', low_memory=False)
df27 = pd.read_csv(data_folder + r'\CRMLSSold202602.csv', low_memory=False)
df28 = pd.read_csv(data_folder + r'\CRMLSSold202603.csv', low_memory=False)
df29 = pd.read_csv(data_folder + r'\CRMLSSold202604.csv', low_memory=False)
df30 = pd.read_csv(data_folder + r'\CRMLSSold202605.csv', low_memory=False)

df_raw = pd.concat([df1, df2, df3, df4, df5, df6, df7, df8, df9, df10,
                    df11, df12, df13, df14, df15, df16, df17, df18, df19, df20,
                    df21, df22, df23, df24, df25, df26, df27, df28, df29, df30],
                    ignore_index=True)

df = df_raw[
    (df_raw['PropertyType'] == 'Residential') &
    (df_raw['PropertySubType'] == 'SingleFamilyResidence')
].copy()

print(f'Total rows before filter: {len(df_raw):,}')
print(f'Total rows after filter: {len(df):,}')

Total rows after filter: 794,271
Total rows after filter: 399,157


## 3. Select Features
I'm selecting only features that are available for both on-market and off-market properties. Per guidance, ListPrice and OriginalListPrice are excluded because they introduce target leakage and won't be available for off-market properties.

In [5]:
features = [
    'ClosePrice',       # target variable
    'LivingArea',       # interior square footage
    'BedroomsTotal',    # number of bedrooms
    'BathroomsTotalInteger',  # number of bathrooms
    'LotSizeAcres',     # lot size
    'YearBuilt',        # year home was built
    'GarageSpaces',     # number of garage spaces
    'PoolPrivateYN',    # whether home has a pool
    'Latitude',         # geographic location
    'Longitude',        # geographic location
    'CloseDate'         # needed for train/test split
]

df = df[features].copy()
print(f'Columns selected: {df.columns.tolist()}')
print(f'Shape: {df.shape}')

Columns selected: ['ClosePrice', 'LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger', 'LotSizeAcres', 'YearBuilt', 'GarageSpaces', 'PoolPrivateYN', 'Latitude', 'Longitude', 'CloseDate']
Shape: (399157, 11)


## 4. Remove Rows with Invalid ClosePrice
Only keeping rows where ClosePrice exists and is greater than zero since that's my target variable.

In [6]:
before = len(df)
df = df[df['ClosePrice'].notna() & (df['ClosePrice'] > 0)]
print(f'Rows removed due to invalid ClosePrice: {before - len(df):,}')
print(f'Rows remaining: {len(df):,}')

Rows removed due to invalid ClosePrice: 3
Rows remaining: 399,154


## 5. Handle Missing Values
Since missing values are under 2% across all key columns (confirmed in Week 2 EDA), I'm dropping rows with any missing values. This keeps the data clean and avoids introducing artificial values through imputation.

In [7]:
before = len(df)
df = df.dropna()
print(f'Rows dropped due to missing values: {before - len(df):,}')
print(f'Rows remaining: {len(df):,}')

Rows dropped due to missing values: 59,028
Rows remaining: 340,126


## 6. Encode Categorical Columns
PoolPrivateYN is a True/False column — converting it to 1/0 so the model can use it numerically.

In [8]:
df['PoolPrivateYN'] = df['PoolPrivateYN'].map({True: 1, False: 0, 'True': 1, 'False': 0})

print('PoolPrivateYN value counts:')
print(df['PoolPrivateYN'].value_counts())

PoolPrivateYN value counts:
PoolPrivateYN
0    284079
1     56047
Name: count, dtype: int64


## 7. Add Property Age Feature
Instead of using YearBuilt directly, I'm engineering a more meaningful feature: how old the home was at the time of sale.

In [9]:
df['CloseDate'] = pd.to_datetime(df['CloseDate'], errors='coerce')
df['PropertyAge'] = df['CloseDate'].dt.year - df['YearBuilt']

# drop any rows where PropertyAge came out negative or unrealistic
df = df[df['PropertyAge'] >= 0]

print(df['PropertyAge'].describe().round(1))

count    340095.0
mean         47.6
std          27.0
min           0.0
25%          26.0
50%          47.0
75%          68.0
max         249.0
Name: PropertyAge, dtype: float64


## 8. Train/Test Split
Per instructions:
- **Test set** = most recent full month available = May 2026
- **Training set** = 12 months immediately before that (May 2025 – April 2026)

I'm starting with 12 months as my training window and will experiment with different window sizes in later weeks to find the optimal value.

In [10]:
df['YearMonth'] = df['CloseDate'].dt.to_period('M')

# test set = May 2026
test = df[df['YearMonth'] == '2026-05'].copy()

# training set = 12 months before test (May 2025 to April 2026)
train = df[
    (df['YearMonth'] >= '2025-05') &
    (df['YearMonth'] <= '2026-04')
].copy()

print(f'Training set: {len(train):,} rows')
print(f'Test set:     {len(test):,} rows')
print(f'Training window: {train["YearMonth"].min()} to {train["YearMonth"].max()}')
print(f'Test window:     {test["YearMonth"].min()} to {test["YearMonth"].max()}')

Training set: 113,754 rows
Test set:     10,412 rows
Training window: 2025-05 to 2026-04
Test window:     2026-05 to 2026-05


## 9. Define Features and Target
Dropping CloseDate and YearMonth since they were only needed for the split — not actual model inputs.

In [11]:
feature_cols = [
    'LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger',
    'LotSizeAcres', 'PropertyAge', 'GarageSpaces',
    'PoolPrivateYN', 'Latitude', 'Longitude'
]

X_train = train[feature_cols]
y_train = train['ClosePrice']

X_test = test[feature_cols]
y_test = test['ClosePrice']

print(f'X_train shape: {X_train.shape}')
print(f'X_test shape:  {X_test.shape}')

X_train shape: (113754, 9)
X_test shape:  (10412, 9)


## 10. Save Cleaned Data
Saving the full cleaned dataset and the train/test splits as CSVs for use in Week 4 modeling.

In [12]:
output_folder = r'C:\Users\monik\OneDrive - University of Illinois - Urbana\Desktop\IDX Exchange_DS\outputs'

train[feature_cols + ['ClosePrice']].to_csv(output_folder + r'\train.csv', index=False)
test[feature_cols + ['ClosePrice']].to_csv(output_folder + r'\test.csv', index=False)

print('train.csv and test.csv saved to outputs folder!')

train.csv and test.csv saved to outputs folder!


## 11. Preprocessing Summary

**Decisions made:**
- Excluded `ListPrice` and `OriginalListPrice` to avoid target leakage and ensure the model works for off-market properties
- Dropped rows with missing values — missing data was under 2% so minimal data loss
- Encoded `PoolPrivateYN` as 1/0
- Engineered `PropertyAge` from `YearBuilt` and `CloseDate`
- Test set = May 2026 (most recent full month)
- Training window = 12 months (May 2025 – April 2026) — will experiment with other window sizes in later weeks

### Why so many rows dropped?
In Week 2 each column individually had under 2% missing values. However when dropping 
rows, a row is removed if ANY of its columns has a missing value. With 9 feature columns, 
these small percentages stack up across columns, resulting in 15% of rows being removed. 
We still have 340,126 clean rows remaining which is more than enough for modeling.